client/X11/xf_gdi.c

```diff
@@ -970,96 +970,86 @@ static BOOL xf_gdi_surface_update_frame(xfContext* xfc, UINT16 tx, UINT16 ty,
 	return ret;
 }
 
+static BOOL xf_gdi_update_screen(xfContext* xfc,
+                                 const SURFACE_BITS_COMMAND* cmd,
+                                 const BYTE* pSrcData)
+{
+	BOOL ret;
+	XImage* image;
+
+	if (!xfc || !pSrcData)
+		return FALSE;
+
+	XSetFunction(xfc->display, xfc->gc, GXcopy);
+	XSetFillStyle(xfc->display, xfc->gc, FillSolid);
+	image = XCreateImage(xfc->display, xfc->visual, xfc->depth, ZPixmap, 0,
+	                     (char*) pSrcData, cmd->width, cmd->height, xfc->scanline_pad, 0);
+	XPutImage(xfc->display, xfc->primary, xfc->gc, image, 0, 0,
+	          cmd->destLeft, cmd->destTop, cmd->width, cmd->height);
+	XFree(image);
+	ret = xf_gdi_surface_update_frame(xfc, cmd->destLeft, cmd->destTop, cmd->width,
+	                                  cmd->height);
+	XSetClipMask(xfc->display, xfc->gc, None);
+	return ret;
+}
+
 static BOOL xf_gdi_surface_bits(rdpContext* context,
                                 const SURFACE_BITS_COMMAND* cmd)
 {
-	XImage* image;
 	BYTE* pSrcData;
-	BYTE* pDstData;
 	xfContext* xfc = (xfContext*) context;
-	BOOL ret = TRUE;
+	BOOL ret = FALSE;
+	DWORD format;
 	rdpGdi* gdi = context->gdi;
-	REGION16 invalidRegion;
 	xf_lock_x11(xfc, FALSE);
 
 	switch (cmd->codecID)
 	{
 		case RDP_CODEC_ID_REMOTEFX:
+			format = PIXEL_FORMAT_BGRX32;
+
 			if (!rfx_process_message(context->codecs->rfx, cmd->bitmapData,
-			                         PIXEL_FORMAT_XRGB32, cmd->bitmapDataLength,
+			                         format, cmd->bitmapDataLength,
 			                         cmd->destLeft, cmd->destTop,
 			                         gdi->primary_buffer, gdi->dstFormat, gdi->stride,
```

Comment:
@realjiangms will update the pull tomorrow, late here. as for the disadvantage, you can of course always use a pointer directly to the XImage->data element.